src/core.js

```diff
@@ -2740,13 +2763,14 @@ Strophe.Connection = class Connection {
      */
     _sasl_challenge_cb (elem) {
       const challenge = atob(Strophe.getText(elem));
-      const response = this._sasl_mechanism.onChallenge(this, challenge);
-      const stanza = $build('response', {'xmlns': Strophe.NS.SASL});
-      if (response !== "") {
-        stanza.t(btoa(response));
-      }
-      this.send(stanza.tree());
-      return true;
+      this._sasl_mechanism.onChallenge(this, challenge).then(response => {;
+            const stanza = $build('response', {'xmlns': Strophe.NS.SASL});
+            if (response !== "") {
+              stanza.t(btoa(response));
+            }
+            this.send(stanza.tree());
+            return true;
+       });
```

Comment:
Using `await` would be a bit nicer here.

```
const response = await this._sasl_mechanism.onChallenge(this, challenge);
```

Then the rest of the code can stay as it was and the diff gets a bit smaller.